lib/srv/desktop/rdp/rdpclient/src/client.rs

```diff
@@ -349,6 +368,69 @@ impl Client {
         write_stream.write_all(&resp).await?;
         Ok(())
     }
+
+    async fn write_rdpdr(
+        write_stream: &mut RdpWriteStream,
+        x224_processor: Arc<Mutex<X224Processor>>,
+        pdu: RdpdrPdu,
+    ) -> ClientResult<()> {
+        // Process the RDPDR PDU.
+        let encoded = Client::x224_process_svc_messages(
+            x224_processor,
+            SvcProcessorMessages::<Rdpdr>::new(vec![SvcMessage::from(pdu)]),
+        )
+        .await?;
+
+        // Write the RDPDR PDU to the RDP server.
+        write_stream.write_all(&encoded).await?;
+        Ok(())
+    }
+
+    /// Processes an x224 frame on a blocking thread.
+    ///
+    /// We use a blocking task here so we don't block the tokio runtime
+    /// while waiting for the `x224_processor` lock, or while processing the frame.
+    /// This function ensures `x224_processor` is locked only for the necessary duration
+    /// of the function call.
+    async fn x224_process(
+        x224_processor: Arc<Mutex<X224Processor>>,
+        frame: BytesMut,
+    ) -> SessionResult<Vec<u8>> {
+        global::TOKIO_RT
+            .spawn_blocking(move || {
+                x224_processor
+                    .lock()
+                    .map_err(|err| reason_err!("x224_processor.lock()", "PoisonError: {:?}", err))?
```

Comment:
Consider extracting function to reuse it in `x224_process_svc_messages`
```rust
fn lock(x224_processor: &Arc<Mutex<Processor>>) -> Result<MutexGuard<Processor>, SessionError> {
    x224_processor
        .lock()
        .map_err(|err| reason_err!("x224_processor.lock()", "PoisonError: {:?}", err))
}
```

Reply:
ed79c67d8e57eec981eb78cfdcde73f1fd351b1a